### PR TITLE
Fix DatePicker Time Input

### PR DESCRIPTION
**Pull request by chirokas**
Closes # 

## 📝 Description

Closing the Date Picker while the Month and Year Pickers are open causes the Time Input to disappear when reopened.

https://github.com/nextui-org/nextui/blob/1ba2d029a291f5d40cbf03f66107eb2ebed09066/packages/components/date-picker/src/date-picker.tsx#L41-L42

## ⛳️ Current behavior (updates)

https://github.com/nextui-org/nextui/assets/157580465/42a64362-4464-491c-bb35-20a4d6bab7f6

https://github.com/nextui-org/nextui/assets/157580465/8b615ff8-dfa9-484f-b33a-109f49c147be

> Please describe the current behavior that you are modifying

## 🚀 New behavior

https://github.com/nextui-org/nextui/assets/157580465/c9b2c95e-8ac1-41d2-afe9-0950440f8a07

> Please describe the behavior or changes this PR adds

## Summary by CodeRabbit

- **New Features**
  - Enhanced DatePicker with Month and Year Picker functionality.
  - Added `onClose` function to manage date picker closing behavior.

- **Bug Fixes**
  - Improved state management for calendar header expansion.

- **Tests**
  - Introduced tests for Month and Year Picker in DatePicker.
  - Verified controlled and uncontrolled modes for month and year picker.
  - Ensured correct rendering of `CalendarBottomContent`.

### CODE REVIEW

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. please add some description on the issue and what you are trying to fix
2. please add a changeset since you've changed the src code
3. please add tests to cover the logic you added

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. please add tests to `packages/components/date-picker/__tests__/date-picker.test.tsx` to cover the logic you added.

**Review by wingkwong — CHANGES_REQUESTED** [review]
The test cases you added just cover the base case of month and year picker, which is ok. However, I didn't see the logic where the PR is trying to solve in the test case, i.e. calendarBottomContent should be still here after open the month / year picker and the date picker.

**Comment by chirokas** [comment]
If there are any problems, please guide me. Thanks!

File: packages/components/date-picker/__tests__/date-picker.test.tsx [hunk 1]
```diff
@@ -458,4 +458,122 @@ describe("DatePicker", () => {
...
   });
+
+  describe("Month and Year Picker", () => {
+    const onHeaderExpandedChangeSpy = jest.fn();
+
+    afterEach(() => {
+      onHeaderExpandedChangeSpy.mockClear();
+    });
+
+    it("should show the month and year picker (uncontrolled)", async () => {
```
Comment (wingkwong) [inline packages/components/date-picker/__tests__/date-picker.test.tsx:469]: `async` seems unnecessary here. (also for other tests under this describe group)

**Review by wingkwong — APPROVED** [review]

File: packages/components/date-picker/__tests__/date-picker.test.tsx [hunk 2]
```diff
@@ -459,6 +459,124 @@ describe("DatePicker", () => {
...
+
+      triggerPress(button);
+
+      dialog = getByRole("dialog");
+      calendarBottomContent = getByTestId("calendar-bottom-content");
+
+      expect(dialog).toBeVisible();
+      expect(calendarBottomContent).toBeVisible();
+    });
+
```
Comment (jrgarciadev) [inline packages/components/date-picker/__tests__/date-picker.test.tsx:579]:
```suggestion
   });
```

**Pull request merged**